plaidml/core/core.h

```diff
@@ -3,14 +3,51 @@
 #pragma once
 
 #include <cstring>
+#include <exception>
 #include <functional>
 #include <memory>
+#include <sstream>
 #include <stdexcept>
 #include <string>
 #include <vector>
 
 #include "plaidml/core/ffi.h"
 
+#if __has_include(<source_location>)
+#include <source_location>
+using edsl_source_location = std::source_location;
+#elif __cplusplus >= 201402 && __has_include(<experimental/source_location>)
+#include <experimental/source_location>
+using edsl_source_location = std::experimental::source_location;
+#else
+struct edsl_source_location {
+  static edsl_source_location current() { return edsl_source_location{}; }
+  constexpr uint_least32_t line() const noexcept { return 0; }
+  constexpr uint_least32_t column() const noexcept { return 0; }
+  constexpr const char* file_name() const noexcept { return "??"; }
+  constexpr const char* function_name() const noexcept { return "??"; }
+};
+#endif
+
+class ffi_exception : public std::exception {
+  std::string msg_;
+  edsl_source_location loc;
+
+ public:
+  explicit ffi_exception(const std::string& message, edsl_source_location loc) : loc(loc) {
+    std::stringstream ss;
+    if (!strstr(loc.file_name(), "edsl/edsl.h")) {
```

Comment:
Avoid using `strstr` in C++ code, you can use `std::string::find` as an alternative.

Reply:
Ah my bad; I'd removed that but must've copied it back in during my `Exception` experimentation